src/NServiceBus.Core/Unicast/RunningEndpointInstance.cs

```diff
@@ -21,74 +21,116 @@ public RunningEndpointInstance(SettingsHolder settings, HostingComponent hosting
 
         public async Task Stop()
         {
-            if (stopped)
+            if (status == Status.Stopped)
             {
                 return;
             }
 
             try
             {
+                // Ensures to only continue if all parallel invocations can rely on the endpoint instance to be fully stopped.
                 await stopSemaphore.WaitAsync().ConfigureAwait(false);
 
-                if (stopped)
+                if (status >= Status.Stopping) // Another invocation is already handling Stop
                 {
                     return;
                 }
 
-                Log.Info("Initiating shutdown.");
+                status = Status.Stopping;
 
-                // Cannot throw by design
-                await receiveComponent.Stop().ConfigureAwait(false);
-                await featureComponent.Stop().ConfigureAwait(false);
-                // Can throw
-                await transportInfrastructure.Stop().ConfigureAwait(false);
-            }
-            catch (Exception exception)
-            {
-                Log.Warn("Exception occurred during shutdown of the transport.", exception);
+                try
+                {
+                    Log.Info("Initiating shutdown.");
+
+                    // Cannot throw by design
+                    await receiveComponent.Stop().ConfigureAwait(false);
+                    await featureComponent.Stop().ConfigureAwait(false);
+
+                    // Can throw
+                    await transportInfrastructure.Stop().ConfigureAwait(false);
+                }
+                catch (Exception exception)
+                {
+                    Log.Error("Shutdown of the transport infrastructure failed.", exception);
+                    
+                    // TODO: Not throwing because reason unknown :)
```

Comment:
@andreasohlund I think it is good that we log here but I've increased the log level to ERROR because this indicate potential incomplete shutdown. Also, shouldn't we just throw here too? Unless there is a good reason but then we should have that added as a comment.